### ISSUE TITLE
Graphical Corruption on Mobile (iOS)

### ISSUE THREAD

**Issue by thejustinwalsh** [HeapsIO/heaps#602]
Experiencing graphical corruption on an iPhone Xs when rendering 3D objects in both the World Sample and the Skin sample, and potentially others however I only tested those two thus far.

I've attached screen recordings below.

World iOS native target:
![world-app-ios](https://user-images.githubusercontent.com/72912/56428891-d671f900-6275-11e9-89ab-9973f0922e7b.gif)

World JS target:
![world-web-ios mov](https://user-images.githubusercontent.com/72912/56428901-deca3400-6275-11e9-8f86-0b7c72f63e77.gif)

**Comment by thejustinwalsh** [comment]
I have merged in #599 and am experiencing the same results as above vs head.

**Comment by ncannasse** [comment]
We made some fixes wrt GpuParticles recently.

Could you confirm that:
- disabling Gpu Particles fixes the problem
- recent changes in GpuParticles fixes the problem

**Comment by ncannasse** [comment]
Closing due to lack of answer. Assuming fixed.

**Comment by NewKrok** [comment]
Hey, I just checked it with my phone (iPhone 6s) - https://heaps.io/samples/world.html - And it's the same. I was not sure how old is this build there so I built it from 1.7.0 and it's the same there.
http://flashplusplus.net/haxe/world3d/index.html

In my own project it's the same too and when I dont change the camera's y position than it looks it works.
Can it be a problem with the projectedPosition calculation in the BaseMesh shader? I have weird issues on desktop browser too and when I set the z manually than it helps but it's impossible to maintain (I will try to investigate it but my knowledge is limited in this topic).

![image](https://user-images.githubusercontent.com/13141660/70867926-4bb7da00-1f7b-11ea-8e2a-35fcb0f42bda.png)

**Comment by ncannasse** [comment]
This might be related to z buffer precision given we are using a very far z
plane. Does changing s3d.camera.zFar work better?

**Comment by NewKrok** [comment]
The default settings in the sample is:
`s3d.camera.zNear = 1;`
`s3d.camera.zFar = 100;`

I modified the demo, you can check it too here:
http://flashplusplus.net/haxe/world3d-zfar/index.html
(When I modify the zFar/zNear and I'm using a cameracontroller it has no any affect - maybe it was my fault - so I just added a dynamic cam position)

For me it looks like more or less the same
![image](https://user-images.githubusercontent.com/13141660/70903646-97ac6280-1fff-11ea-958d-fb67389fdda3.png)

**Comment by ncannasse** [comment]
Did you try with latest GIT ?

**Comment by NewKrok** [comment]
Hm good point, I have just tried with **1.7.0** (Haxe 4.0.0)

Sadly it's the same with the latest version too, and it's the same in iOS Safari and is iOS Chrome.
http://flashplusplus.net/haxe/world3d-git/index.html
![image](https://user-images.githubusercontent.com/13141660/70941644-28f5f600-204d-11ea-9155-7e658490cd79.png)
![image](https://user-images.githubusercontent.com/13141660/70941678-3a3f0280-204d-11ea-9acb-0337e1bb4b67.png)
![image](https://user-images.githubusercontent.com/13141660/70941416-b258f880-204c-11ea-873e-14b760cca6e0.png)

**Comment by ncannasse** [comment]
Interesting... Does adding a single tree and a single rock produces same
results ?

**Comment by NewKrok** [comment]
With a single Rock is the same:
http://flashplusplus.net/haxe/world3d-single_rock/index.html
![image](https://user-images.githubusercontent.com/13141660/70943156-7a53b480-2050-11ea-88a1-90bd53603385.png)

With a single Tree is the same:
http://flashplusplus.net/haxe/world3d-single_tree/index.html
![image](https://user-images.githubusercontent.com/13141660/70943169-83dd1c80-2050-11ea-8a1e-de00b895f29c.png)

With a single Cube is different BUT strange too. 2 sides from 6 looks noisy with a simple blue texture:
http://flashplusplus.net/haxe/world3d-single_cube/index.html
![image](https://user-images.githubusercontent.com/13141660/70943755-df5bda00-2051-11ea-8af2-816a48d4b2db.png)


Btw I just tried without "snow" and it's the same too.
http://flashplusplus.net/haxe/world3d-no_snow/index.html

**Comment by ncannasse** [comment]
That's weird indeed. Can't pinpoint from just the screenshots what's going
wrong.
Does the normal cube sample (Base3D) work as expected?
What about Base2D ?

**Comment by NewKrok** [comment]
Base2d looks ok, but Base3d is similar (Sorry for the picture quality :D):
Can I add some extra log which can helps for you?

![IMG_1219](https://user-images.githubusercontent.com/13141660/70951401-96148600-2063-11ea-9781-81741c675b54.jpg)
![IMG_1222](https://user-images.githubusercontent.com/13141660/70951402-97de4980-2063-11ea-804e-17ef68de4fb5.jpg)

**Comment by ncannasse** [comment]
It's good that we get something reproducible with Base3D, as it's much more simple example.

Can you confirm that it's some kind of flickering that is similar to the original issue poster's video ?
Are you trying in JS (in browser) and/or HL native ?

**Comment by NewKrok** [comment]
Yes it’s some kind of flickering. With base3d sample it’s more rare, and it looks it’s mostly happening with some specific rotation. 
I have tried it just with js build and it’s happening just in a mobile browser.

I have some similar issue with chrome browser with PC, maybe it’s fully different topic but maybe it could be useful for you too:
I have 2 meshes with grid primtives.
They z positions are almost the same.
They have 2 different textures.
When the camera is close than it looks ok.
When you zoom out they are randomly “flickering”

**Comment by CrazyFlasher** [comment]
Hi. Just took latest Heaps from Git and have the same issue on Samsung Galaxy A41
![heaps](https://user-images.githubusercontent.com/1607138/96716841-42bf4280-13ae-11eb-80fe-bf3f280505b5.gif)

**Comment by Rohit-Ghadi** [comment]
I think is issue in texture loader.
use ImageBitmapLoader instead of TextureLoader

**Comment by amrezzd** [comment]
Ok it's working on `x86` build very well. However `armeabi-v7a` device has the issue I've mentioned above. I'm not sure if I'm missing something. Here is my cmake file.
[CMakeLists.txt](https://github.com/HeapsIO/heaps/files/8110751/CMakeLists.txt)
